Add "Toil Management and Operational Improvements" section to SRE Maturity

mmazur opened the pull request:
It replaces the existing "Room to improve" one since there was quite an overlap there. Phase 2 and 3 still remain as mostly stubs, though expanded ones. (More info will be available to fill them out with once a few internal SRE teams gather some more experience on the subject.)

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign  after the PR has been reviewed.
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2FSRE).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/operate-first/SRE/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

anishasthana commented on dashboards/README.md:
```diff
@@ -0,0 +1,18 @@
+## Grafana Monitoring Dashboards
+
+[Grafana](https://grafana.com/oss/grafana) is a multi-platform open source analytics and interactive visualization tool. It allows you to query, create dashboards, explore and alert on metrics when connected to supported data sources.
```
`explore, and alert`

hemajv replied:
accepted the `oxford comma` :)

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/operate-first/SRE/pull/21#pullrequestreview-529462144" title="Approved">anishasthana</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign  after the PR has been reviewed.
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2FSRE).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/operate-first/SRE/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

HumairAK commented on dashboards/README.md:
```diff
@@ -0,0 +1,18 @@
+These monitoring dashboards are useful for visualizing the metrics (such as Prometheus metrics) collected from the applications to track and analyze the system's overall health.
+The dashboards can be altered, saved and reused as per your specific monitoring needs. 
+
+You will find the exported JSON file for each of the dashboards in this folder.
```
fyi as far as I know we'll be using the dashboard custom resource that's provided by the grafana operator
see [here](https://github.com/integr8ly/grafana-operator/blob/master/documentation/dashboards.md) for more info. I think you can just paste the raw json in the custom resource though.

hemajv replied:
updated :+1:

sesheta commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/operate-first/SRE/pull/21#pullrequestreview-529462144" title="Approved">anishasthana</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2FSRE).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/operate-first/SRE/blob/master/OWNERS)~~ [anishasthana]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

mmazur committed:
Add "Toil Management and Operational Improvements" section

It replaces the existing "Room to improve" one since there was quite an
overlap there. Phase 2 and 3 still remain as mostly stubs, though
expanded ones. (More info will be available to fill them out with once
a few internal SRE teams gather some more experience on the subject.)

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+There are five requirements that need to be met for your SRE team to start burning down toil and improving its operational efficiency:
+1. Oncall time must be limited to at most 50% of total engineering time.
+2. The remaining time must be dedicated to work that contributes to operational improvements – toil reduction, automation, architectural improvements, etc.
+3. Engineers must have an understanding of what toil is, how to spot it in day-to-day work and what are the ways of reducing it.
```
Engineers and SRE Managers must have an ...

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
-It isn't necessary to formally measure how much time is spent on keeping things running vs. feature development or automation.
-That can come later, however you are familiar with the concept of Toil.
+There are five requirements that need to be met for your SRE team to start burning down toil and improving its operational efficiency:
+1. Oncall time must be limited to at most 50% of total engineering time.
```
Management supports a policy such as Oncall time must be ...

wking commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+* Contact with upstream (whichever team owns the product you’re SRE-ing) is well-established and they have resources reserved to work on issues raised by your SRE team. Workarounds for production issues applied by SREs are always temporary, while upstream works on a proper fix.
+* Toil items are tracked and a simple scoring system is used to prioritize work.
+* Regular meetings are held to review recent on-call shifts and ensure all improvement opportunities have been captured and scored to prioritize future work.
+* 
```
nit: unused list entry.  Drop this line?

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+#### Toil Bankruptcy and the Dangers of Unmanaged Toil
 
-### Room to improve - Phase 3: The 'Run' Phase
+Should an SRE team fail at reducing toil faster than it accumulates due to service growth, such a team can find itself in a very serious predicament – spending all of its time on operations tasks with zero room to improve (e.g. automate) anything.
```
s/due to service growth//

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+Should an SRE team fail at reducing toil faster than it accumulates due to service growth, such a team can find itself in a very serious predicament – spending all of its time on operations tasks with zero room to improve (e.g. automate) anything.
 
-*This is a stub. Please help by adding content. Ref:[SIGSRE-87](https://issues.redhat.com/browse/SIGSRE-87)*
+The chart below visualizes what happens to an SRE team's time commitments when their system is growing 20% quarter over quarter yet they do not address the toily part of operations (marked in red).
```
Can you please add an attribution + context to where this data came from?

mmazur replied:
I've created this chart with a simple [sheet](https://docs.google.com/spreadsheets/d/1x3THIRbH1cCtlZAi6KMaRV_VW53jbtrBnxlq4B8Uyrs/edit#gid=0), but since it's in our Red Hat sheets, I can't link to that directly.

Hm, what I could do is add a short separate document with the raw data in a table and an explanation of how the calculations were derived and link to that.

mmazur replied:
Added the detailed explanation. (Linked at the end of that paragraph).

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+*This is a stub. Please help by adding content. Ref: [SIGSRE-86](https://issues.redhat.com/browse/SIGSRE-86)*
+
+* Contact with upstream (whichever team owns the product you’re SRE-ing) is well-established and they have resources reserved to work on issues raised by your SRE team. Workarounds for production issues applied by SREs are always temporary, while upstream works on a proper fix.
+* Toil items are tracked and a simple scoring system is used to prioritize work.
```
I think readers are going to want more out of this bullet. Some even initial/starting proposal for how to do this.

mmazur replied:
I amended the wording somewhat here, but I can't really give fully fledged suggestions yet.

Some of the SRE teams I've talked to have only fairly recently dabbled with scoring their work on the toil aspects of it. What I was hoping to do is leave the walk and run section in this minimal state for Q3 and then revisit the talks with those teams starting Q4, summarize their experiences and use that to fill out the walk and run sections.

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+
+### Toil Management and Operational Improvements - Phase 3: The 'Run' Phase
+
+*This is a stub. Please help by adding content. Ref: [SIGSRE-87](https://issues.redhat.com/browse/SIGSRE-87)*
```
remove this line

mmazur replied:
I did, but the 'stub' line is still present in the "Walk" phase section. Both are like 3 bullet points, so I think at least some indication that these are still unfinished would be warranted.

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+
+* Contact with upstream (whichever team owns the product you’re SRE-ing) is well-established and they have resources reserved to work on issues raised by your SRE team. Workarounds for production issues applied by SREs are always temporary, while upstream works on a proper fix.
+* Toil items are tracked and a simple scoring system is used to prioritize work.
+* Regular meetings are held to review recent on-call shifts and ensure all improvement opportunities have been captured and scored to prioritize future work.
```
Maybe clarify here that development work to reduce toil should be tracked through existing planning ceremonies?

mmazur replied:
Rewrote the 3 bullet points to make it clearer that the goal is to: review ops regularly, always card the outcomes, then prio those cards during standard work planning.

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+
+*This is a stub. Please help by adding content. Ref: [SIGSRE-87](https://issues.redhat.com/browse/SIGSRE-87)*
+
+* Cooperation with upstream has matured to a point where high priority toil reduction items raised by the SRE team get immediate attention simply because SRE needs the toil reduced as soon as possible.
```
s/ simply because SRE needs the toil reduced as soon as possible.//

jeremyeder commented on sre_maturity.md:
```diff
@@ -175,27 +175,68 @@ In this phase you do ad-hoc security assessments, scans and pentesting.
+
+* Cooperation with upstream has matured to a point where high priority toil reduction items raised by the SRE team get immediate attention simply because SRE needs the toil reduced as soon as possible.
+  * This is separate from customer-visible incidents, where upstream’s quick involvement would stem from a different escalation path.
+* Toil reduction opportunities are scored with a multi-variable scoring system tailored to a particular SRE team’s goals and priorities.
```
Examples for this TBD in blog format?

mmazur replied:
Yup, wait for Q4, then write up a working system teams have been using for a while (as a blog post) and link to that.

mmazur committed:
Address some of the comments

mmazur committed:
At detailed explanation of the chart
